addons/sale_stock/models/sale_order.py

```diff
@@ -155,17 +156,21 @@ def _compute_picking_ids(self):
         for order in self:
             order.delivery_count = len(order.picking_ids)
 
-    @api.depends('user_id', 'company_id')
+    @api.depends('user_id', 'company_id', 'order_line')
     def _compute_warehouse_id(self):
         for order in self:
-            default_warehouse_id = self.env['ir.default'].with_company(
-                order.company_id.id)._get_model_defaults('sale.order').get('warehouse_id')
-            if order.state in ['draft', 'sent'] or not order.ids:
-                # Should expect empty
-                if default_warehouse_id is not None:
-                    order.warehouse_id = default_warehouse_id
-                else:
-                    order.warehouse_id = order.user_id.with_company(order.company_id.id)._get_default_warehouse_id()
+            is_delivery_set = any(line._is_delivery() for line in order.order_line)
+            if is_delivery_set and order.carrier_id.delivery_type == 'onsite' and order.carrier_id.warehouse_id:
+                order.warehouse_id = order.carrier_id.warehouse_id
+            else:
+                default_warehouse_id = self.env['ir.default'].with_company(
+                    order.company_id.id)._get_model_defaults('sale.order').get('warehouse_id')
+                if order.state in ['draft', 'sent'] or not order.ids:
+                    # Should expect empty
+                    if default_warehouse_id is not None:
+                        order.warehouse_id = default_warehouse_id
+                    else:
+                        order.warehouse_id = order.user_id.with_company(order.company_id.id)._get_default_warehouse_id()
```

Comment:
could be cleaned up visually:

use `continue` after `order.warehouse_id = order.carrier_id.warehouse_id`, then we can avoid the else + indents after that